<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />






















<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=6.4.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.4.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.4.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.4.0">


  <link rel="mask-icon" href="/images/logo.svg?v=6.4.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '6.4.0',
    sidebar: {"position":"left","display":"hide","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="snakemake是一个用来编写任务流程的工具，用python写的，因此其执行的流程脚本也比较通俗易懂，易于理解，可以看做用户友好版的make。 其实流程控制是复杂任务（在生信领域很常见）必需的关注点。只是snakemake对于代码功力不够的人来说，在写好代码与重复流程的花销的trade off上，还不如一遍遍重复流程。。但是真的是一个写好了就很好用的东西。 12# installingconda">
<meta name="keywords" content="生物信息,snakemake">
<meta property="og:type" content="article">
<meta property="og:title" content="snakemake实践">
<meta property="og:url" content="http://yoursite.com/2019/02/20/snakemake/index.html">
<meta property="og:site_name" content="Qizhengyang">
<meta property="og:description" content="snakemake是一个用来编写任务流程的工具，用python写的，因此其执行的流程脚本也比较通俗易懂，易于理解，可以看做用户友好版的make。 其实流程控制是复杂任务（在生信领域很常见）必需的关注点。只是snakemake对于代码功力不够的人来说，在写好代码与重复流程的花销的trade off上，还不如一遍遍重复流程。。但是真的是一个写好了就很好用的东西。 12# installingconda">
<meta property="og:locale" content="en">
<meta property="og:image" content="http://yoursite.com/images/1550217959613.png">
<meta property="og:image" content="http://yoursite.com/images/1550218038489.png">
<meta property="og:image" content="http://yoursite.com/images/dag1.png">
<meta property="og:image" content="http://yoursite.com/images/1550648786426.png">
<meta property="og:image" content="http://yoursite.com/images/1550648837730.png">
<meta property="og:image" content="http://yoursite.com/images/1550648869840.png">
<meta property="og:updated_time" content="2019-02-20T08:50:31.480Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="snakemake实践">
<meta name="twitter:description" content="snakemake是一个用来编写任务流程的工具，用python写的，因此其执行的流程脚本也比较通俗易懂，易于理解，可以看做用户友好版的make。 其实流程控制是复杂任务（在生信领域很常见）必需的关注点。只是snakemake对于代码功力不够的人来说，在写好代码与重复流程的花销的trade off上，还不如一遍遍重复流程。。但是真的是一个写好了就很好用的东西。 12# installingconda">
<meta name="twitter:image" content="http://yoursite.com/images/1550217959613.png">






  <link rel="canonical" href="http://yoursite.com/2019/02/20/snakemake/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>snakemake实践 | Qizhengyang</title>
  









  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Qizhengyang</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="/about/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/tags/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/archives/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />Search</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/20/snakemake/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Qizhengyang">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">snakemake实践
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-20 12:00:57 / Modified: 16:50:31" itemprop="dateCreated datePublished" datetime="2019-02-20T12:00:57+08:00">2019-02-20</time>
            

            
              

              
            
          </span>

          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>snakemake是一个用来编写任务流程的工具，用python写的，因此其执行的流程脚本也比较通俗易懂，易于理解，可以看做用户友好版的make。</p>
<p>其实流程控制是复杂任务（在生信领域很常见）必需的关注点。只是snakemake对于代码功力不够的人来说，在写好代码与重复流程的花销的trade off上，还不如一遍遍重复流程。。但是真的是一个写好了就很好用的东西。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> installing</span><br><span class="line">conda install -c bioconda snakemake</span><br></pre></td></tr></table></figure>
<h2 id="简单的例子"><a href="#简单的例子" class="headerlink" title="简单的例子"></a>简单的例子</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">cd $HOME</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Create a folder where we will run our commands:</span><br><span class="line">mkdir snakemake-example</span><br><span class="line">cd snakemake-example</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Make a fake genome:</span><br><span class="line">touch genome.fa</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Make some fake data:</span><br><span class="line">mkdir fastq</span><br><span class="line">touch fastq/Sample1.R1.fastq.gz fastq/Sample1.R2.fastq.gz</span><br><span class="line">touch fastq/Sample2.R1.fastq.gz fastq/Sample2.R2.fastq.gz</span><br></pre></td></tr></table></figure>
<p>snakemake 脚本</p>
<figure class="highlight makefile"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">SAMPLES = ['Sample1', 'Sample2']</span><br><span class="line"></span><br><span class="line">rule all:</span><br><span class="line">    input:</span><br><span class="line">        expand('&#123;sample&#125;.txt', sample=SAMPLES)</span><br><span class="line"></span><br><span class="line">rule quantify_genes:</span><br><span class="line">    input:</span><br><span class="line">        genome = 'genome.fa',</span><br><span class="line">        r1 = 'fastq/&#123;sample&#125;.R1.fastq.gz',</span><br><span class="line">        r2 = 'fastq/&#123;sample&#125;.R2.fastq.gz'</span><br><span class="line">    output:</span><br><span class="line">        '&#123;sample&#125;.txt'</span><br><span class="line">    shell:</span><br><span class="line">        'echo &#123;input.genome&#125; &#123;input.r1&#125; &#123;input.r2&#125; &gt; &#123;output&#125;'</span><br></pre></td></tr></table></figure>
<p>运行</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> 学院集群上运行</span><br><span class="line">snakemake --snakefile Snakefile</span><br></pre></td></tr></table></figure>
<p><img src="/images/1550217959613.png" alt="1550217959613"></p>
<p><img src="/images/1550218038489.png" alt="1550218038489"></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake --forceall --dag | dot -Tpng &gt; dag1.png</span><br></pre></td></tr></table></figure>
<p><img src="/images/dag1.png" alt="dag1"></p>
<h2 id="snakemake规则"><a href="#snakemake规则" class="headerlink" title="snakemake规则"></a>snakemake规则</h2><ul>
<li>Snakemake基于规则执行命令，规则一般由input, output,shell三部分组成。除了rule all，其余必须有output</li>
<li>Snakemake可以自动确定不同规则的输入输出的依赖关系，根据<strong>时间戳</strong>来判断文件是否需要重新生成</li>
<li>Snakemake以{sample}.fa形式进行<strong>文件名通配</strong>，用{wildcards.sample}获取sample的实际文件名</li>
<li>Snakemake用expand()生成多个文件名，本质是Python的列表推导式</li>
<li>Snakemake可以在规则外直接写Python代码，在规则内的run里也可以写Python代码。</li>
<li>Snakefile的第一个规则通常是rule all，根据all里的文件决定执行哪些rule。如上面的例子，注释掉all里的input则不执行第二条rule，<strong>（推断未尝试：rule all里定义最终的输出文件，程序也能执行，那么rule里的输出文件在什么时候会被删除，是在所有rule运行完之后，还是在判断出该输出文件不会被用到的时候？）</strong></li>
<li>在output中的结果文件可以是未存在目录中的文件,这时会自动创建不存在的目录（不需要事先建文件夹，这个功能实在是方便）</li>
</ul>
<h2 id="snakemake命令"><a href="#snakemake命令" class="headerlink" title="snakemake命令"></a>snakemake命令</h2><ol>
<li><p>wildcards: 用来获取通配符匹配到的部分，例如对于通配符”{dataset}/file.{group}.txt”匹配到文件101/file.A.txt，则{wildcards.dataset}就是101，{wildcards.group}就是A。</p>
</li>
<li><p>temp: 通过temp方法可以在所有rule运行完后删除指定的中间文件，eg.output: temp(“f1.bam”)。</p>
</li>
<li><p>protected: 用来指定某些<strong>中间文件是需要保留的</strong>，eg.output: protected(“f1.bam”)。</p>
</li>
</ol>
<h2 id="snakemake执行"><a href="#snakemake执行" class="headerlink" title="snakemake执行"></a>snakemake执行</h2><p>一般讲所有的参数配置写入Snakefile后直接在Snakefile所在路径执行snakemake命令即可开始执行流程任务，如果只有一个snakefile的话，连文件都不用写。一些常用的参数：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">--snakefile, -s 指定Snakefile，否则是当前目录下的Snakefile</span><br><span class="line">--dryrun, -n  不真正执行，一般用来查看Snakefile是否有错</span><br><span class="line">--printshellcmds, -p   输出要执行的shell命令</span><br><span class="line">--reason, -r  输出每条rule执行的原因,默认FALSE</span><br><span class="line">--cores, --jobs, -j  指定运行的核数，若不指定，则使用最大的核数</span><br><span class="line">--force, -f 重新运行第一条rule或指定的rule</span><br><span class="line">--forceall, -F 重新运行所有的rule，不管是否已经有输出结果</span><br><span class="line">--forcerun, -R 重新执行Snakefile，当更新了rule时候使用此命令</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span>一些可视化命令</span><br><span class="line"><span class="meta">$</span> snakemake --dag | dot -Tpdf &gt; dag.pdf</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span>集群投递</span><br><span class="line">snakemake --cluster "qsub -V -cwd -q 节点队列" -j 10</span><br><span class="line"><span class="meta">#</span> --cluster: 集群运行指令</span><br><span class="line"><span class="meta">#</span> qusb -V -cwd -q， 表示输出当前环境变量(-V),在当前目录下运行(-cwd), 投递到指定的队列(-q), 如果不指定则使用任何可用队列</span><br><span class="line"><span class="meta">#</span> --local-cores N: 在每个集群中最多并行N核</span><br><span class="line"><span class="meta">#</span> --cluster-config/-u FILE: 集群配置文件</span><br></pre></td></tr></table></figure>
<h2 id="自己写的snakemake-file"><a href="#自己写的snakemake-file" class="headerlink" title="自己写的snakemake file"></a>自己写的snakemake file</h2><figure class="highlight makefile"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 2019-2-19 10:28:52 part</span></span><br><span class="line"><span class="comment"># GATK snakemake</span></span><br><span class="line"><span class="comment"># qizhengyang</span></span><br><span class="line"></span><br><span class="line">from os.path import join</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">GENOME = 'genome/HWB.chromosome.fa'</span><br><span class="line">GTF = 'genes/HWB.gene.models.gtf'</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">(SAMPLES,) = glob_wildcards('pairedDIR/&#123;sample&#125;_1P.fq.gz')</span><br><span class="line">PATTERN_R1 = join('pairedDIR', '&#123;sample&#125;_1P.fq.gz')</span><br><span class="line">PATTERN_R2 = join('pairedDIR', '&#123;sample&#125;_2P.fq.gz')</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">rule all:</span><br><span class="line">    input:</span><br><span class="line">        'star_index_2pass/',</span><br><span class="line">        expand('star_1pass/&#123;sample&#125;SJ.out.tab', sample=SAMPLES),</span><br><span class="line">        'star_index_2pass/',</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;Aligned.out.sam', sample=SAMPLES),</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;_rg_added_sorted.bam', sample=SAMPLES),</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;_dedup.bam', sample=SAMPLES),</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;_dedup_split.bam', sample=SAMPLES),</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;.vcf', sample=SAMPLES),</span><br><span class="line">        expand('star_2pass/&#123;sample&#125;_filtered.vcf', sample=SAMPLES)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">rule star_index:</span><br><span class="line">    input:</span><br><span class="line">        genome = GENOME,</span><br><span class="line">        gtf = GTF</span><br><span class="line">    output:</span><br><span class="line">        star_index = 'star_index/'</span><br><span class="line"></span><br><span class="line">    log:</span><br><span class="line">        'star_index.log'</span><br><span class="line">    threads: 8</span><br><span class="line">    run:</span><br><span class="line">        <span class="comment"># star 1-pass index, OK</span></span><br><span class="line">        shell('STAR --runThreadN &#123;threads&#125; --runMode genomeGenerate'</span><br><span class="line">              ' --genomeDir &#123;output.star_index&#125;'</span><br><span class="line">              ' --genomeFastaFiles &#123;input.genome&#125;'</span><br><span class="line">              ' --sjdbGTFfile &#123;input.gtf&#125;'</span><br><span class="line">              ' 2&gt; &#123;log&#125;')</span><br><span class="line"></span><br><span class="line">rule star_1pass_align:</span><br><span class="line">    input:</span><br><span class="line">        index = 'star_index/',</span><br><span class="line">        r1 = PATTERN_R1,</span><br><span class="line">        r2 = PATTERN_R2</span><br><span class="line">    output:</span><br><span class="line">        index = 'star_1pass/&#123;sample&#125;SJ.out.tab'</span><br><span class="line">    threads: 8</span><br><span class="line">    params:</span><br><span class="line">        prefix = './star_1pass/&#123;sample&#125;'</span><br><span class="line">    <span class="comment"># 在使用params之前是报错的，NameError,The name 'sample' is unknown in this context</span></span><br><span class="line">    run:</span><br><span class="line">        <span class="comment"># star 1-pass align, OK</span></span><br><span class="line">        shell('STAR --runThreadN &#123;threads&#125; --genomeDir &#123;input.index&#125;'</span><br><span class="line">              ' --readFilesIn &#123;input.r1&#125; &#123;input.r2&#125;'</span><br><span class="line">              ' --readFilesCommand zcat'</span><br><span class="line">              ' --outFileNamePrefix &#123;params.prefix&#125;')</span><br><span class="line"></span><br><span class="line">rule star_2pass_index:</span><br><span class="line">    input:</span><br><span class="line">        genome = GENOME,</span><br><span class="line"></span><br><span class="line">        <span class="comment"># 这里必需加expand，不然会报错：Wildcards in input files cannot be determined from output files:</span></span><br><span class="line">        <span class="comment"># 'sample'。</span></span><br><span class="line">        <span class="comment"># 很奇怪，确实是需要所有样本的剪接位点信息，我之前没有注意到。。还好报错，不然就尴尬了</span></span><br><span class="line">        <span class="comment"># 然后用--sjdbFileChrStartEnd参数将所有样品的SJ.out.tab文件作为输入的annotated junction进行第二次建index</span></span><br><span class="line">        <span class="comment"># http://www.bioinfo-scrounger.com/archives/288</span></span><br><span class="line"></span><br><span class="line">        <span class="comment"># 这里不能加三个引号（"""或'''注释）</span></span><br><span class="line"></span><br><span class="line">        splice_site = expand('star_1pass/&#123;sample&#125;SJ.out.tab', sample=SAMPLES)</span><br><span class="line">    output:</span><br><span class="line">        index = 'star_index_2pass/'</span><br><span class="line">    threads: 8</span><br><span class="line">    run:</span><br><span class="line">        <span class="comment"># star 2-pass index, OK</span></span><br><span class="line">        shell('STAR --runThreadN &#123;threads&#125; --runMode genomeGenerate'</span><br><span class="line">              ' --genomeDir &#123;output.index&#125;'</span><br><span class="line">              ' --genomeFastaFiles &#123;input.genome&#125;'</span><br><span class="line">              ' --sjdbFileChrStartEnd &#123;input.splice_site&#125;')</span><br><span class="line"></span><br><span class="line">rule star_2pass_align:</span><br><span class="line">    input:</span><br><span class="line">        index = 'star_index_2pass/',</span><br><span class="line">        r1 = PATTERN_R1,</span><br><span class="line">        r2 = PATTERN_R2</span><br><span class="line">    output:</span><br><span class="line">        sam = 'star_2pass/&#123;sample&#125;Aligned.out.sam'</span><br><span class="line">    threads: 8</span><br><span class="line">    params:</span><br><span class="line">        prefix = 'star_2pass/&#123;sample&#125;'</span><br><span class="line">    run:</span><br><span class="line">        <span class="comment"># star 2-pass align</span></span><br><span class="line">        shell('STAR --runThreadN &#123;threads&#125; --genomeDir &#123;input.index&#125;'</span><br><span class="line">              ' --readFilesIn &#123;input.r1&#125; &#123;input.r2&#125;'</span><br><span class="line">              ' --readFilesCommand zcat'</span><br><span class="line">              ' --outFileNamePrefix &#123;params.prefix&#125;')</span><br><span class="line"></span><br><span class="line">rule picard:</span><br><span class="line">    input:</span><br><span class="line">        sam = 'star_2pass/&#123;sample&#125;Aligned.out.sam'</span><br><span class="line">    output:</span><br><span class="line">        bam = 'star_2pass/&#123;sample&#125;_rg_added_sorted.bam'</span><br><span class="line">    run:</span><br><span class="line">        <span class="comment"># RGID和RGSM的sample必须是&#123;wildcards.sample&#125;，不然</span></span><br><span class="line">        <span class="comment"># The name 'sample' is unknown in this context. Please make sure that you defined that variable.</span></span><br><span class="line">        <span class="comment"># Also note that braces not used for variable access have to be escaped by repeating them, i.e. &#123;&#123;print $1&#125;&#125;</span></span><br><span class="line">        shell('picard AddOrReplaceReadGroups'</span><br><span class="line">              ' I=&#123;input.sam&#125;'</span><br><span class="line">              ' O=&#123;output.bam&#125;'</span><br><span class="line">              ' SO=coordinate'</span><br><span class="line">              ' RGID=&#123;wildcards.sample&#125;'</span><br><span class="line">              ' RGLB=rna'</span><br><span class="line">              ' RGPL=illumina'</span><br><span class="line">              ' RGPU=hiseq'</span><br><span class="line">              ' RGSM=&#123;wildcards.sample&#125;')</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">rule picard_markduplicates:</span><br><span class="line">    input:</span><br><span class="line">        bam = 'star_2pass/&#123;sample&#125;_rg_added_sorted.bam'</span><br><span class="line">    output:</span><br><span class="line">        dedup_bam = 'star_2pass/&#123;sample&#125;_dedup.bam'</span><br><span class="line">    params:</span><br><span class="line">        dedup_metrices = 'star_2pass/&#123;sample&#125;_dedup.metrics'</span><br><span class="line">    run:</span><br><span class="line">        shell('picard MarkDuplicates'</span><br><span class="line">              ' I=&#123;input.bam&#125;'</span><br><span class="line">              ' O=&#123;output.dedup_bam&#125;'</span><br><span class="line">              ' CREATE_INDEX=true'</span><br><span class="line">              ' VALIDATION_STRINGENCY=SILENT'</span><br><span class="line">              ' M=&#123;params.dedup_metrices&#125;')</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">rule gatk_split:</span><br><span class="line">    input:</span><br><span class="line">        dedup_bam = 'star_2pass/&#123;sample&#125;_dedup.bam',</span><br><span class="line">        genome = GENOME</span><br><span class="line">    output:</span><br><span class="line">        split_bam = 'star_2pass/&#123;sample&#125;_dedup_split.bam'</span><br><span class="line">    run:</span><br><span class="line">        shell('java -jar $GATK -T SplitNCigarReads'</span><br><span class="line">              ' -R &#123;input.genome&#125;'</span><br><span class="line">              ' -I &#123;input.dedup_bam&#125;'</span><br><span class="line">              ' -o &#123;output.split_bam&#125;'</span><br><span class="line">              ' -rf ReassignOneMappingQuality'</span><br><span class="line">              ' -RMQF 255'</span><br><span class="line">              ' -RMQT 60'</span><br><span class="line">              ' -U ALLOW_N_CIGAR_READS')</span><br><span class="line">rule gatk:</span><br><span class="line">    input:</span><br><span class="line">        bam = 'star_2pass/&#123;sample&#125;_dedup_split.bam',</span><br><span class="line">        genome = GENOME</span><br><span class="line">    output:</span><br><span class="line">        vcf = 'star_2pass/&#123;sample&#125;.vcf'</span><br><span class="line">    run:</span><br><span class="line">        shell('java -jar $GATK -T HaplotypeCaller'</span><br><span class="line">              ' -R &#123;input.genome&#125;'</span><br><span class="line">              ' -I &#123;input.bam&#125;'</span><br><span class="line">              ' -dontUseSoftClippedBases'</span><br><span class="line">              ' -stand_call_conf 20.0'</span><br><span class="line">              ' -o &#123;output.vcf&#125;')</span><br><span class="line"></span><br><span class="line">rule gatk_filter:</span><br><span class="line">    input:</span><br><span class="line">        genome = GENOME,</span><br><span class="line">        vcf = 'star_2pass/&#123;sample&#125;.vcf'</span><br><span class="line">    output:</span><br><span class="line">        'star_2pass/&#123;sample&#125;_filtered.vcf'</span><br><span class="line">    run:</span><br><span class="line">        shell('java -jar $GATK'</span><br><span class="line">              ' -T VariantFiltration'</span><br><span class="line">              ' -R &#123;input.genome&#125;'</span><br><span class="line">              ' -V &#123;input.vcf&#125;'</span><br><span class="line">              ' -window 35'</span><br><span class="line">              ' -cluster 3'</span><br><span class="line">              ' -filterName FS -filter <span class="string">"FS &gt; 30.0"</span>'</span><br><span class="line">              ' -filterName QD -filter <span class="string">"QD &lt; 2.0"</span>'</span><br><span class="line">              ' -o &#123;output&#125;')</span><br></pre></td></tr></table></figure>
<h2 id="集群上运行"><a href="#集群上运行" class="headerlink" title="集群上运行"></a>集群上运行</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> 查看节点状态</span><br><span class="line">pestat</span><br><span class="line"><span class="meta">#</span> 测试 dry run</span><br><span class="line">snakemake -n</span><br><span class="line"><span class="meta">#</span> 检测某条rule</span><br><span class="line">snakemake -n -r star_index</span><br><span class="line"><span class="meta">#</span> 可视化</span><br><span class="line"><span class="meta">#</span> snakemake --dag | dot -Tpdf &gt; dag.pdf</span><br><span class="line">snakemake --forceall --dag | dot -Tpdf &gt; dag.pdf</span><br><span class="line"><span class="meta">#</span># 投递任务 -cwd不能用 [qsub: illegal -c value]</span><br><span class="line"><span class="meta">#</span> snakemake --cluster "qsub -V -cwd -q high" -j 20</span><br><span class="line"></span><br><span class="line">snakemake --cluster "qsub -V -q high" -j 20</span><br><span class="line"><span class="meta">#</span> 报错</span><br><span class="line"><span class="meta">#</span> Complete log: /home02/qizhengyang/qizhengyang/gatk_rna/.snakemake/log/2019-02-20T152054.823542.snakemake.log</span><br><span class="line"><span class="meta">#</span> less snakejob.star_index.254.sh.e980069 查看错误信息</span><br><span class="line"><span class="meta">#</span>  Output directories must be flagged with directory()，</span><br><span class="line"><span class="meta">#</span> 集群上做的时候输出目录要用 directory() ，在实验室服务器上没有遇上这种情况</span><br><span class="line"></span><br><span class="line">snakemake -s Snakefile1 --cluster "qsub -V -q high" -j 20</span><br><span class="line"><span class="meta">#</span> -j 必须指定，否则Error: you need to specify the maximum number of jobs to be queued or executed at the same time with --jobs.</span><br><span class="line"><span class="meta">#</span> Here, -j denotes the number of jobs submitted being submitted to the cluster at the same time</span><br><span class="line"><span class="meta">#</span> 没有重新运行star_index这个任务</span><br></pre></td></tr></table></figure>
<p>第一次</p>
<p><img src="/images/1550648786426.png" alt="1550648786426"></p>
<p>第二次</p>
<p><img src="/images/1550648837730.png" alt="1550648837730"></p>
<p>snakemake会自动判断哪些任务可以并行处理：</p>
<p><img src="/images/1550648869840.png" alt="1550648869840"></p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ol>
<li><a href="https://snakemake.readthedocs.io/en/latest/executable.html?highlight=cluster" target="_blank" rel="noopener">snakemake docs</a></li>
<li><a href="https://github.com/slowkow/snakefiles/pull/1/files" target="_blank" rel="noopener">RNAseq variant calling pipeline </a></li>
<li><a href="https://slowkow.com/notes/snakemake-tutorial/#1-installing-snakemake" target="_blank" rel="noopener">Build bioinformatics pipelines with Snakemake</a></li>
<li><a href="https://github.com/slowkow/snakefiles/blob/master/kallisto/Snakefile" target="_blank" rel="noopener">kallisto-snakefiles</a></li>
<li><a href="http://pedagogix-tagc.univ-mrs.fr/courses/ABD/practical/snakemake/snake_intro.html" target="_blank" rel="noopener">Writing a RNA-Seq workflow with snakemake</a></li>
<li><a href="https://www.cmwonderland.com/2018/05/05/snakemake/" target="_blank" rel="noopener">snakemake初步</a></li>
<li><a href="https://www.jianshu.com/p/b400dc7c5eea" target="_blank" rel="noopener">RNA-seq 检测变异之 GATK 最佳实践流程</a></li>
<li><a href="https://github.com/leipzig/snakemake-example/blob/master/Snakefile" target="_blank" rel="noopener">snakemake-example</a></li>
<li><a href="https://www.jianshu.com/p/8e57fd2b81b2" target="_blank" rel="noopener">我最喜欢的流程管理工具</a></li>
</ol>

      
    </div>

    

    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/生物信息/" rel="tag"># 生物信息</a>
          
            <a href="/tags/snakemake/" rel="tag"># snakemake</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/02/11/机器停止/" rel="next" title="the machine stops">
                <i class="fa fa-chevron-left"></i> the machine stops
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name"></p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">24</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">9</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          

          

          
          

          
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#简单的例子"><span class="nav-text">简单的例子</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#snakemake规则"><span class="nav-text">snakemake规则</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#snakemake命令"><span class="nav-text">snakemake命令</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#snakemake执行"><span class="nav-text">snakemake执行</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#自己写的snakemake-file"><span class="nav-text">自己写的snakemake file</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#集群上运行"><span class="nav-text">集群上运行</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder"></span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> v3.7.1</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme – <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Mist</a> v6.4.0</div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    
	
    

    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=6.4.0"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=6.4.0"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=6.4.0"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=6.4.0"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=6.4.0"></script>



  



  










  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

  

</body>
</html>
